Bluetooth: Support setting long periodic advertising data

rugeGerritsen opened the pull request:
If the data is longer than what can fit in one fragment, the data is sent to the controller in multiple operations.

rugeGerritsen committed:
Bluetooth: HCI: Max periodic adv data length is 252 bytes

Use the define instead

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -462,20 +492,30 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
 			/* We can't necessarily set one AD field in a single step. */
 			while (offset < data[j].data_len) {
-				buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
+				BUILD_ASSERT(sizeof(struct bt_hci_cp_le_set_per_adv_data)
+					     == UINT8_MAX);
+				BUILD_ASSERT(sizeof(struct bt_hci_cp_le_set_ext_adv_data)
+					     == UINT8_MAX);
+				BUILD_ASSERT(sizeof(struct bt_hci_cp_le_set_ext_scan_rsp_data)
+					     == UINT8_MAX);
```
These asserts may conflict with what is being requested here: https://github.com/zephyrproject-rtos/zephyr/issues/37710

rugeGerritsen replied:
I'm not sure if I follow. The size of these commands are defined by the spec. If we want to optimize the communication between cores, then we should rather modify the length we pass to `net_buf_add()`

Thalley replied:
The point is that 255 is the maximum size of the commands, and that always sending 255 octets (which we currently do) even if we just set a single octet of data is inefficient. Of course we could keep the structs at 255 octets and, as you suggest, simply modify the number of octets we add to the buffer.

rugeGerritsen replied:
Yeah. There are multiple ways of solving it. Lets wait for some more input on this

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -444,15 +444,45 @@ static int hci_set_adv_ext_complete(struct bt_le_ext_adv *adv, uint16_t hci_op,
 }
 
-static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op,
+static int hci_set_adv_ext_fragmented(const struct bt_le_ext_adv *adv, uint16_t hci_op,
```
Smart move to use this for periodic advertising data too, but I do wonder if the name is still fitting. Doesn't necessarily need a change, but something to consider.

carlescufi replied:
No matter what the resolution is here, I consider the current implementation unacceptable as it is. Please take out the common parts into a common function that can be then used by the per-command functions like we do everywhere else, including hci.c.

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -462,20 +492,30 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
+					     == UINT8_MAX);
+				BUILD_ASSERT(sizeof(struct bt_hci_cp_le_set_ext_scan_rsp_data)
+					     == UINT8_MAX);
+				buf = bt_hci_cmd_create(hci_op, UINT8_MAX);
```
Perhaps a `MAX` of the `sizeof` the 3 structs would be better to use than `UINT8_MAX`. Alternative create a lookup size by `hci_op`.

rugeGerritsen replied:
If you're ok with it, I would just keep it as is. It is just a matter of style. The correctness is ensured by the build assertions

Thalley replied:
I have a slight preference to using MAX and sizeof (which also adds some semantic reasoning behind the value), but with the BUILD_ASSERTs I'm ok with this.

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -484,39 +524,43 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
-					offset += set_data->len - 2;
+					set_data[data_offset + 0] = len + 1;
+					set_data[data_offset + 1] = type;
+					memcpy(&set_data[data_offset + 2], data[j].data,
```
```suggestion
					(void)memcpy(&set_data[data_offset + 2], data[j].data,
```

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -484,39 +524,43 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
-					set_data->data[1] = type;
-					memcpy(&set_data->data[2], data[j].data, set_data->len - 2);
-					offset += set_data->len - 2;
+					set_data[data_offset + 0] = len + 1;
+					set_data[data_offset + 1] = type;
+					memcpy(&set_data[data_offset + 2], data[j].data,
+					       set_data[len_offset] - 2);
```
Given that you already started using `_offset` variables, consider adding variables `data_len_offset`, `data_type_offset` and `data_value_offset`

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -484,39 +524,43 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
-					memcpy(&set_data->data[0], &data[j].data[offset],
-					       set_data->len);
-					offset += set_data->len;
+					memcpy(&set_data[data_offset], &data[j].data[offset],
```
```suggestion
					(void)memcpy(&set_data[data_offset], &data[j].data[offset],
```

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -462,20 +492,28 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
+				size_t cmd_size =
+					MAX(sizeof(struct bt_hci_cp_le_set_per_adv_data),
+					    MAX(sizeof(struct bt_hci_cp_le_set_ext_adv_data),
+						sizeof(struct bt_hci_cp_le_set_ext_scan_rsp_data)));
+				buf = bt_hci_cmd_create(hci_op, cmd_size);
```
I think you need an empty line between declarations and the rest of the code

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -462,20 +492,28 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
 			/* We can't necessarily set one AD field in a single step. */
 			while (offset < data[j].data_len) {
-				buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
+				size_t cmd_size =
```
```suggestion
				const size_t cmd_size =
```

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -484,47 +522,54 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
-					set_data->data[1] = type;
-					memcpy(&set_data->data[2], data[j].data, set_data->len - 2);
-					offset += set_data->len - 2;
+					size_t data_len_offset = data_offset;
+					size_t data_type_offset = data_offset + 1;
+					size_t data_value_offset = data_offset + 2;
```
```suggestion
					const size_t data_len_offset = data_offset;
					const size_t data_type_offset = data_offset + 1;
					const size_t data_value_offset = data_offset + 2;
```

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1610,13 +1655,41 @@ int bt_le_per_adv_set_param(struct bt_le_ext_adv *adv,
+	set_data = net_buf_add(buf, sizeof(*set_data));
+	(void)memset(set_data, 0, sizeof(*set_data));
+
+	err = set_data_add_complete(set_data->data, BT_HCI_LE_PER_ADV_FRAG_MAX_LEN,
+				    ad, ad_len,
+				    &set_data->len);
```
```suggestion
	err = set_data_add_complete(set_data->data, BT_HCI_LE_PER_ADV_FRAG_MAX_LEN,
				    ad, ad_len, &set_data->len);
```

Thalley reviewed (approved):
Minor code style comments, otherwise LGTM

alwa-nordic commented on subsys/bluetooth/host/adv.c:
```diff
@@ -520,14 +550,18 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
+					set_data[data_len_offset] = len + 1;
+					set_data[data_type_offset] = type;
+					(void)memcpy(&set_data[data_value_offset], data[j].data,
+						     set_data[data_type_offset] - 2);
```
Typo here? Should it be `len_offset` instead of `data_type_offset`?

rugeGerritsen replied:
Yes, you are right

jhedberg commented on subsys/bluetooth/host/adv.c:
```diff
@@ -444,15 +444,45 @@ static int hci_set_adv_ext_complete(struct bt_le_ext_adv *adv, uint16_t hci_op,
+	size_t len_offset;
+	size_t data_offset;
+
+	if (hci_op == BT_HCI_OP_LE_SET_PER_ADV_DATA) {
+		handle_offset = offsetof(struct bt_hci_cp_le_set_per_adv_data, handle);
+		op_offset = offsetof(struct bt_hci_cp_le_set_per_adv_data, op);
+		fragment_pref_offset = 0; /* Unused, not present. */
+		len_offset = offsetof(struct bt_hci_cp_le_set_per_adv_data, len);
+		data_offset = offsetof(struct bt_hci_cp_le_set_per_adv_data, data);
+	} else {
+		handle_offset = offsetof(struct bt_hci_cp_le_set_ext_adv_data, handle);
+		op_offset = offsetof(struct bt_hci_cp_le_set_ext_adv_data, op);
+		fragment_pref_offset = offsetof(struct bt_hci_cp_le_set_ext_adv_data, frag_pref);
+		len_offset = offsetof(struct bt_hci_cp_le_set_ext_adv_data, len);
+		data_offset = offsetof(struct bt_hci_cp_le_set_ext_adv_data, data);
+
+		BUILD_ASSERT(offsetof(struct bt_hci_cp_le_set_ext_adv_data, handle)
+			     == offsetof(struct bt_hci_cp_le_set_ext_scan_rsp_data, handle));
+		BUILD_ASSERT(offsetof(struct bt_hci_cp_le_set_ext_adv_data, op)
+			     == offsetof(struct bt_hci_cp_le_set_ext_scan_rsp_data, op));
+		BUILD_ASSERT(offsetof(struct bt_hci_cp_le_set_ext_adv_data, frag_pref)
+			     == offsetof(struct bt_hci_cp_le_set_ext_scan_rsp_data, frag_pref));
+		BUILD_ASSERT(offsetof(struct bt_hci_cp_le_set_ext_adv_data, len)
+			     == offsetof(struct bt_hci_cp_le_set_ext_scan_rsp_data, len));
+		BUILD_ASSERT(offsetof(struct bt_hci_cp_le_set_ext_adv_data, data)
+			     == offsetof(struct bt_hci_cp_le_set_ext_scan_rsp_data, data));
+	}
```
This really doesn't look very nice. So you're overloading a single function to deal with three different (but sufficiently similar) structs, by treating them as byte-arrays and having magic offsets to access data. It kind of defeats the entire purpose why the stack uses packed structs since those exist specifically to avoid byte array-based access and magic offsets.

I don't have a finalised alternative solution to avoid this, but I'm hoping to start a discussion/brainstorming for how to do this better. Would it e.g. make sense to factor out the common part (dealing with the actual payload) to a separate helper function, and then having HCI command (and therefore struct) specific higher level functions calling that helper function?

Thalley replied:
I agree that this isn't particularly pretty. Separating common functionality into functions sounds like a good idea, and then have specific functions to deal with each of the different structs/operations. 

@jhedberg How would you feel about something like

```
handle_offset = get_handle_offset(hci_op):
op_offset = get_op_offset(hci_op);
...
```

Such that we split getting the different offsets into multiple functions? 
It would create a couple new, but smaller, functions, but it will still use the offset values that you didn't like so much.

jhedberg replied:
That's slightly better, but it still doesn't eliminate the offsets and treating the structs as byte arrays.

rugeGerritsen replied:
@Thalley, @jhedberg, where do we want to put the discussion?
I've extracted out some notes here:

Currently the host supports configuring the controller with long advertising data.
It was pointed out in https://github.com/zephyrproject-rtos/zephyr/pull/40280 that some refactoring should be done to make the code more maintainable. The refactoring should make it easier to implement support for setting long periodic advertising data.

The current implementation sends the data to the controller one AD structure at the time.
If the AD structure is too long, the data is forwarded in two operations.

Things that makes the implementation complex: 
 - One AD structure cannot be set in one operation as the maximum length AD structure length is 255 + 1 byte.
   The maximum length that can be passed to the controller is 251.
 - The AD data is provided as a two-dimensional array.
   The application provides one array of AD structures, the host adds one more dimension so that it can add the local name.
 - Determining the operation parameter (first/intermediate/last) is hard to read when considering the facts above:
   It needs to know both the offset, and the iterator values for both dimensions.
 - The complete local name can be shortened.
 - We are not allowed to pass zero-length data commands to the controller.

Possible ways of simplifying the implementation:
 - First put all advertising data into a scratch buffer.
   This will require more RAM, but will allow us to simplify the code a lot.
 - Flatten the two-dimensional array before sending the data to the controller.
 - Limit the maximum AD structure length to 240 bytes so that we don't need to split it.
 - Never shorten the complete local name when setting long advertising data.

jhedberg replied:
@rugeGerritsen all sound like potentially reasonable options to me, each with its own pros and cons. It'd be good to get input from other host developers here.

trond-snekvik replied:
I haven't followed the discussion, but Rubin's summary sounds pretty good to me. A couple of thoughts and questions:
- Should the normal advertising data also not shorten the name, for consistency?
- Will this increase the amount of copying being done on the advertisement data? The mesh protocol sends all its traffic through the (legacy) advertiser, so excessive copying could have a non-trivial impact on performance.
- I'm sure someone will have a use case for 250 byte AD data in the future. We shouldn't make this 240 byte limitation if we end up painting ourselves into a corner 
- Can the assembly buffer be allocated in the controller memory (or as an HCI packet) to reduce copying and memory consumption?

Either way, I don't have any objections.

rugeGerritsen replied:
@trond-snekvik , some answers and additional nots to your thoughts and questions:

- We want to set all the data in a single operation (HCI command) if possible as it is not allowed by the spec to set the data in multiple operations while the advertiser is running. Therefore, we will only refactor setting _long_ advertising data. That is, the cases where the advertising data cannot fit into a single HCI packet. For mesh applications we always set the data in a single operation.
- The controller is already allocating memory for advertising data reassembly. The host is forwarding HCI packets to this buffer.
- The advertising name is shortened for normal advertising data if the complete local name cannot fit.
- If we use a scratch buffer in the host, there will be more copying. Instead of just copying directly into HCI packets, we will now have one intermediate copy operation. Note: This will only be done when the advertising data does not fit within a single HCI packet. The only benefit with this approach is that it creates a bit cleaner code.
- I agree that we should not put a limitation on the length of AD data.

Thalley replied:
> First put all advertising data into a scratch buffer.
This will require more RAM, but will allow us to simplify the code a lot.

If I understand this correct, this would copy all the data from the array of bt_data supplied by the application (plus the name added by the stack) into an array of max size 1650, right? The additional memory here will thus be the size of this intermediate array, correct? 

> Flatten the two-dimensional array before sending the data to the controller.

This one I don't understand; aren't we already doing that when we transform the array of bt_date (the LTV structures) into a single uint8_t array for the command? 

The way I see it, we can either A) put the entire advertising data from the application into an intermediate array, and then in another step split this array into smaller parts that can be sent to the controller in multiple commands. The alternative B) is that we don't use an intermediate array and copy `X` octets from the application supplied data into the command, and then do that multiple times. The challenge here is of course keeping track of what has already been sent, as the data exists in a 2-dimensional array. 

A) is by far the simplest; not just to implement but also to modify, adapt, reread etc. but at the cost of the RAM of the intermediate array (could possibly just be a stack allocated array). This approach will, besides the additional RAM, also require 2 memcpys: 1 from app data to intermediate array, and 1 from intermediate array to the HCI command structures.
B) is on the other hand "free" in terms of memory as we don't allocate the additional array, and will only require a single memory copy (I guess is what we also just do today, unless I'm mistaken). 

For most use cases A) is fine (especially we can keep the array in the stack, and not require that memory to be static), as the advertising data is fairly rarely set or updated (and especially those large ones). For Mesh it may impact performance significantly, unless we will never require segmentation of the Mesh advertising data. 

Please correct me if I have misunderstood or misremembered some things about how we do it today.

rugeGerritsen replied:
I agree with your thoughts here. The decision making factor is readability vs code size. 

For option A) I would like to allocate the reassembly buffer statically. Allocating a 1650 bytes buffer on a stack is not nice.

rugeGerritsen replied:
Maybe we should really consider to keep the solution present in this PR?

alwa-nordic replied:
@jhedberg rugeGerritsen seems unconvinced. We have to take a hard stance, make a better case or let this PR pass.

alwa-nordic replied:
Some brainstorming:
- Decoupling AD array flattening and HCI command construction and issuing.
  - We could encapsulate the flattening in a stream object. I.e.
    ```
    struct flattened_ad_stream;
    flattened_ad_stream_new(stream, ad_count, ads);
    flattened_ad_stream_read(stream, buf_size, buf);
    flattened_ad_stream_is_empty(stream);
    ```
- Serializing the three similar HCI commands.
  - We could make `set_data` a union of the three packed structs and replace the use of offsets with pointers.
  - The part that uses the above mentioned pointers can be a separate function.

Thalley replied:
@alwa-nordic Wouldn't the `struct flattened_ad_stream` just be a `net_buf_simple` with a `uint8_t array`?

alwa-nordic replied:
@Thalley The stream can be lazy and not use extra memory.

Thalley replied:
Ah, I think I know what you mean now :)

rugeGerritsen replied:
@alwa-nordic , @Thalley , @jhedberg . Should we block this PR until you have sorted out what direction you want to go? 
I would be happy if we could get this merged.

Thalley replied:
Personally I'd rather do the simple implementation now, and then if we see any performance issues, we can look into improving it.

rugeGerritsen replied:
The simple solution = Keep it as is, or rewrite it to use another temporary static array (not thread safe)?

Thalley replied:
I vote for keep it as is.

jhedberg replied:
I don't quite see how the buffering issue is related to the refactoring issue (to treat the command-unique details in helper functions or similar)

Thalley replied:
> I don't quite see how the buffering issue is related to the refactoring issue (to treat the command-unique details in helper functions or similar)

Wow, I had completely forgotten about the refactoring issue... That still needs to be resolved indeed.

rugeGerritsen replied:
@Thalley , do you have some suggestion on what to refactor out? As mentioned in the PR where we implemented support for long advertising data, it is hard to extract something in a nice way. There are just many parameters to take into account

Thalley replied:
@rugeGerritsen I think that @alwa-nordic suggestions in https://github.com/zephyrproject-rtos/zephyr/pull/40280#discussion_r768457697 were pretty sound

cvinayak replied:
Thinking aloud here (ignore me):

I almost every time start with copy paste existing implementation that I feel is close to solving a new software requirement and make necessary (maybe minimal) changes to meet the new functionality. Functionality being of higher precedence compared to optimal implementation.

Have a system test (samples and/or BabbleSim CI) similar to typical application use case to cover (TDD) the implementation so far.

On both self-review and by other reviewers, a question on refactoring will arise, and now there is a chance to refactor out the original and new copy pasted new solution into common reusable code (a new commit in the PR).

I also believe by maintaining some common patterns of code (copy paste inside a file), the compiler does inherently do better code optimization (compared to self designed optimizations which are new patterns seen by the compiler). This is where zephyr.lst file is my friend, providing some insights/clues/evidences to justify refactoring out some reusable code.

Compare flash and RAM usages when refactoring, copy paste is not good for maintenance, hence these blocks be refactored into single re-usable function, such refactoring does not change flash or RAM usage as compiler already inherently would have done the same.

rugeGerritsen replied:
Created https://github.com/zephyrproject-rtos/zephyr/pull/41537

jhedberg commented on subsys/bluetooth/host/adv.c:
```diff
@@ -462,20 +492,29 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
+					    MAX(sizeof(struct bt_hci_cp_le_set_ext_adv_data),
+						sizeof(struct bt_hci_cp_le_set_ext_scan_rsp_data)));
+
+				buf = bt_hci_cmd_create(hci_op, cmd_size);
```
Is the code somewhere overwriting the length field that's part of the HCI command header? The `bt_hci_cmd_create()` function encodes its second parameter into that header field, however at this point it seems the code doesn't yet know what the final length of the command.

rugeGerritsen replied:
Yes, it does. It sets the length to the maximum value (as done for other commands)

jhedberg replied:
I was mainly concerned with the `bt_hci_cp_le_set_per_adv_data` struct, since it's smaller than the two others. However, now that I checked the HCI spec it seems like there might be an off-by-one issue with the data field of this struct (defined as `uint8_t data[251]`. From the HCI spec:
`0 to 252 | The number of octets in the Advertising Data parameter`

If that were to be fixed then all the structs would have the same size.

rugeGerritsen replied:
@jhedberg , that is already fixed as part of this PR :). Now the struct is defined to have to correct maximum data length of 252 bytes.

jhedberg replied:
Ok. I guess I was confused by this construction of MAX statements, since that seems to imply that they are not identical in size.

jhedberg replied:
Anyway, I suppose this entire concern/confusion will go away when the suggested refactoring gets done

rugeGerritsen committed:
Bluetooth: Host: Support setting long periodic adv data

If the advertiser is not running, the host can now set
periodic advertising data in multiple operations.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen commented:
Updated the PR to use `ad_stream` created in https://github.com/zephyrproject-rtos/zephyr/pull/41537

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -659,6 +659,53 @@ static int set_sd(struct bt_le_ext_adv *adv, const struct bt_ad *sd,
+{
+	struct ad_stream stream;
+	struct bt_ad d = { .data = ad, .len = ad_len };
+
+	ad_stream_new(&stream, &d, 1);
+
+	bool is_first_iteration = true;
```
```suggestion
	bool is_first_iteration = true;
	
	ad_stream_new(&stream, &d, 1);

```

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1703,38 +1747,19 @@ int bt_le_per_adv_set_data(const struct bt_le_ext_adv *adv,
+		/* It is not allowed to set periodic advertising data
+		 * in multiple operations while it is running.
+		 */
+		return -EAGAIN;
```
If the data is too long, it doesn't really make sense to tell the caller to try again. 

```suggestion
		return -EINVAL;
```
